### PR TITLE
Fix reflectivity calculation and error propagation in Q grouping in ILL reflectometry reduction

### LINKED ISSUE #22674: ReflectometryILLConvertToQ: problems in Q-grouping
- Comparing to COSMOS, the errors are too large; this seems to be due to wrong error propagation in `ReflectometryILLConvertToQ` which just takes the mean of the grouped errors. The errors should be summed in squares instead.
- Grouping produces distinctively different results in Mantid even when starting from ungrouped data that is almost exactly equal to COSMOS. The reasons should be at least understood.

### PR DESCRIPTION
**Pull request by soininen**
**Description of work.**

This PR reorganizes the `ReflectometryILL*` algorithms such that in the `SumInLambda` mode, the reflected data gets divided by the direct beam only after conversion to *Q* and grouping in *Q*. 

**Report to:** [nobody].

**To test:**

The script below exercises the workflow algorithms both in `SumInLambda` and `SumInQ` modes. The data is available in the unit test data directories. Check that the script runs and the two final workspaces, `reflectivity_sumInLambda` and `reflectivity_sumInQ` look about right; you should see something like this:

reflectivity

```python
# Direct beam first
ReflectometryILLPreprocess('ILL/D17/317369.nxs',
    OutputWorkspace='direct',
    OutputBeamPositionWorkspace='direct_beampos')
ReflectometryILLSumForeground('direct',
    OutputWorkspace='direct_fgd',
    WavelengthRange=[2, 20.])
    
# SumInQ
ReflectometryILLPreprocess('ILL/D17/317370.nxs',
    OutputWorkspace='reflected',
    DirectBeamPositionWorkspace='direct_beampos')
ReflectometryILLSumForeground('reflected',
    OutputWorkspace='reflected_fgd_inQ',
    DirectForegroundWorkspace='direct_fgd',
    SummationType='SumInQ')
ReflectometryILLConvertToQ('reflected_fgd_inQ',
    OutputWorkspace='reflectivity_sumInQ',
    ReflectedBeamWorkspace='reflected',
    DirectBeamWorkspace='direct',
    GroupingQFraction=0.2)

# SumInLambda; same data as above
ReflectometryILLPreprocess('ILL/D17/317370.nxs',
    OutputWorkspace='reflected',
    DirectBeamPositionWorkspace='direct_beampos')
ReflectometryILLSumForeground('reflected',
    OutputWorkspace='reflected_fgd_inLambda',
    DirectForegroundWorkspace='direct_fgd')
ReflectometryILLConvertToQ('reflected_fgd_inLambda',
    OutputWorkspace='reflectivity_sumInLambda',
    ReflectedBeamWorkspace='reflected',
    DirectBeamWorkspace='direct',
    DirectForegroundWorkspace='direct_fgd',
    GroupingQFraction=0.2)
```

Fixes #22674.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by reimundILL** [comment]
- A `DirectForegroundWorkspace` was added to the *`ConvertToQ` algorithm but the diagram does not seem to be updated accordingly
- The `DirectBeamPositionWorkspace` in *`Preprocess` seems to be missing in the diagram of the corresponding documentation

**Review by vreimund — CHANGES_REQUESTED** [review]
I verified these changes by comparing an updated reduction script with the already realized preliminary reduction:
Previous
previous
Now:
pr
This looks good!

But we need to still update the documentation + diagram, see comment above.

**Review by vreimund — CHANGES_REQUESTED** [review]
When running the example given, I expect `XErrors` for the output workspace. This is true for the `SumInQ` option but not for the `SumInLambda` option (`XError` tab missing):
suminlambdamissingresolution

This may require a unit test as well.

My guess: `Divide` does not propagate `Dx`

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryILLConvertToQ.py [hunk 1]
```diff
@@ -148,19 +179,37 @@ def _convertToMomentumTransfer(self, ws):
...
+        """Divide ws by the direct beam."""
+        reflectivityWSName = self._names.withSuffix('reflectivity')
+        reflectivityWS = Divide(LHSWorkspace=ws,
+                                RHSWorkspace=directForegroundWS,
+                                OutputWorkspace=reflectivityWSName,
+                                EnableLogging=self._subalgLogging)
+        self._cleanup.cleanup(ws)
+        self._cleanup.cleanup(directForegroundWS)
+        reflectivityWS.setYUnit('Reflectivity')
+        reflectivityWS.setYUnitLabel('Reflectivity')
```
Comment (reimundILL) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryILLConvertToQ.py]:
```
dxs = ws.readDx(0)
reflectivityWS.setDx(0, dxs)
self._cleanup.cleanup(ws)
```
Seem to solve the problem for me

**Review by vreimund — APPROVED** [review]

**Comment by reimundILL** [comment]
:+1:

**Pull request merged**